omp/reorder/rcm_kernels.cpp

```diff
@@ -0,0 +1,171 @@
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/reorder/rcm_kernels.hpp"
+
+
+#include <algorithm>
+#include <iterator>
+#include <memory>
+#include <queue>
+#include <utility>
+#include <vector>
+
+
+#include <ginkgo/config.hpp>
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/matrix/coo.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/permutation.hpp>
+#include <ginkgo/core/matrix/sparsity_csr.hpp>
+
+
+namespace gko {
+namespace kernels {
+namespace omp {
+/**
+ * @brief The reordering namespace.
+ *
+ * @ingroup reorder
+ */
+namespace rcm {
+
+
+template <typename ValueType, typename IndexType>
+void get_degree_of_nodes(
+    std::shared_ptr<const OmpExecutor> exec,
+    std::shared_ptr<matrix::SparsityCsr<ValueType, IndexType>> adjacency_matrix,
+    std::shared_ptr<gko::Array<IndexType>> node_degrees)
+{
+    auto num_rows = adjacency_matrix->get_size()[0];
+    auto adj_ptrs = adjacency_matrix->get_row_ptrs();
+    auto node_deg = node_degrees->get_data();
+
+#pragma omp parallel for
+    for (auto i = 0; i < num_rows; ++i) {
+        node_deg[i] = adj_ptrs[i + 1] - adj_ptrs[i];
+    }
+}
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+    GKO_DECLARE_RCM_GET_DEGREE_OF_NODES_KERNEL);
+
+
+template <typename IndexType>
+IndexType find_index(std::vector<std::pair<IndexType, IndexType>> &a,
+                     IndexType x)
+{
+    for (auto i = 0; i < a.size(); i++)
+        if (a[i].first == x) return i;
+    return -1;
+}
+
+
+template <typename ValueType, typename IndexType>
+void get_permutation(
+    std::shared_ptr<const OmpExecutor> exec, size_type num_vertices,
+    std::shared_ptr<matrix::SparsityCsr<ValueType, IndexType>> adjacency_matrix,
+    std::shared_ptr<Array<IndexType>> node_degrees,
+    std::shared_ptr<matrix::Permutation<IndexType>> permutation_mat,
+    std::shared_ptr<matrix::Permutation<IndexType>> inv_permutation_mat)
+{
+    IndexType num_vtxs = static_cast<IndexType>(num_vertices);
+    auto adj_ptrs = adjacency_matrix->get_row_ptrs();
+    auto adj_idxs = adjacency_matrix->get_col_idxs();
+    auto node_deg = node_degrees->get_data();
+    auto permutation_arr = permutation_mat->get_permutation();
+    auto inv_permutation_arr = inv_permutation_mat->get_permutation();
+
+    std::queue<IndexType> q;
```

Comment:
A queue is well-suited in principle, but since we only `.pop()` n times from here, we could in principle also use a raw block of memory as in a limited linear queue. 
That would enable us to push to the queue directly, instead of collecting into `to_sort`, follwed by sorting the relevant part of the queue in place.